### ISSUE TITLE
Default attributes for block-support properties don't apply on dynamic blocks

### ISSUE THREAD

**Issue by ryelle** [WordPress/gutenberg#50229]
I'm building a dynamic block, which supports a handful of the built-in style settings (colors, spacing, etc). I can set defaults in `block.json` that apply in the editor, but when I view the front end, it's like nothing is set. `get_block_wrapper_attributes` only returns the block class name, none of the color classes are set, and no `style` attribute.

It appears that the default attributes don't exist in `parsed_block`, which is what `apply_block_supports` uses to render the values in PHP.

Editor, appears correct:

<img width="1259" alt="" src="https://user-images.githubusercontent.com/541093/235543497-3d45c013-579b-4fb5-beff-35a25dbfeb05.png">

Front end, no styles:

<img width="1209" alt="" src="https://user-images.githubusercontent.com/541093/235543496-134cb265-236f-40f1-9abd-fa1eebc0ae11.png">

<details>
<summary>An example block.json</summary>

This sets a default width, colors, and padding, but none of those are applied.

```json
{
	"$schema": "https://schemas.wp.org/trunk/block.json",
	"apiVersion": 2,
	"name": "wporg/local-navigation-bar",
	"title": "Local Navigation Bar",
	"icon": "ellipsis",
	"category": "layout",
	"description": "A special block to handle the local navigation on pages in a section.",
	"textdomain": "wporg",
	"attributes": {
		"align": {
			"type": "string",
			"default": "full"
		},
		"backgroundColor": {
			"type": "string",
			"default": "blueberry-1"
		},
		"textColor": {
			"type": "string",
			"default": "white"
		},
		"style": {
			"type": "object",
			"default": {
				"spacing": {
					"padding": {
						"right": "var:preset|spacing|60",
						"left": "var:preset|spacing|60",
						"top": "16px",
						"bottom": "16px"
					}
				}
			}
		}
	},
	"supports": {
		"align": false,
		"color": {
			"text": true,
			"background": true,
			"link": true
		},
		"spacing": {
			"margin": true,
			"padding": true
		},
	},
	"editorScript": "file:./index.js",
	"editorStyle": "file:./editor-style.css",
	"style": "file:./style.css",
	"viewScript": "file:./view.js"
}
```

</details>

I think this is the same root issue as https://github.com/WordPress/gutenberg/issues/50027, but it affects more than just the alignment.

**Comment by talldan** [comment]
> It appears that the default attributes don't exist in parsed_block, which is what apply_block_supports uses to render the values in PHP.

From what I understand, the default are only usually returned when using an instance of the `WP_Block`:
https://github.com/WordPress/wordpress-develop/blob/1bd815adbbbb48d486a5519ed2a3d34eefe6a407/src/wp-includes/class-wp-block.php#L165-L192

This looks like the code that sets the `$parsed_block`:
https://github.com/WordPress/wordpress-develop/blob/1bd815adbbbb48d486a5519ed2a3d34eefe6a407/src/wp-includes/class-wp-block.php#L252-L263

I don't really know the code well enough to offer suggestions on how to fix it in a backward compatible way. This code has been untouched for years by the looks of it, but @youknowriad last modified it, so might be able to offer suggestions.

**Comment by youknowriad** [comment]
How do you set defaults in block.json? Do you redeclare the style attribute manually? 

---

This issue looks like a symptom of the fact that the block parsing in the server is not done "fully". It's just the grammar parse. Unfortunately, there's no solution for that at the moment as introducing the full HTML parsing is too impactful. 

I'm not really sure why we made sure the default attributes available when using `$WP_Block` instance, this feels like a hacky fix for me but there's a precedent, maybe we can find a way to use that in `apply_block_supports` as well?

**Comment by ryelle** [comment]
> How do you set defaults in block.json? Do you redeclare the style attribute manually?

My block.json is in the issue description. I'm setting some attributes directly, and for spacing, I set `style`, following the docs for [using block supports](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/).

If it helps, for output my render function is simply:

```php
function render( $attributes, $content, $block ) {
	$wrapper_attributes = get_block_wrapper_attributes();
	return sprintf(
		'<div %1$s>%2$s</div>',
		$wrapper_attributes,
		$content
	);
}
```

**Comment by dutchigor** [comment]
I'm encountering the same issue. See a minimal plugin demonstrating the issue here: https://github.com/dutchigor/wordpress-block-supports-render-callback-test

**Comment by colorful-tones** [comment]
I just hit this issue today. I would be keen to see some traction.

**Comment by SantosGuillamot** [comment]
This issue was also reported in this ticket: [here](https://wordpress.org/support/topic/default-block-json-margin-values/). I've been able to reproduce it as well. Any clue/idea how can this be fixed?

**Comment by youknowriad** [comment]
The solution fix can be as I explained here https://github.com/WordPress/gutenberg/issues/50229#issuecomment-1531003958

Make sure to include the default attributes to the "attributes" that are passed to `apply_block_supports` calls. I don't like this fix personally but also don't like the same fix that is present in `WP_Block` instance either (not sure when this was introduced) because it feels like we're hacking the code because we don't have a full "parser" in the server. 

One potential approach as well would be to update `parse_blocks` in the server to always include the default attributes but it can be seen as a small breaking change as well.

**Comment by spencerfinnell** [comment]
I am running in to this as well.

If a block's `style` attribute has been touched/modified the attribute is saved in the block markup, otherwise it's not. Then accessing `$attributes` in the `render_callback` shows a mix of the default `block.json` value (unparsed) and the CSS preset variable that corresponds with the editor setting:

```
// Attribute is unchanged in the editor.
["style"]=>
   array(1) {
    ["spacing"]=>
      array(1) {
        ["blockGap"]=>
          string(7) "0.44rem"
        }
    }
```

```
// Attribute is changed in the editor.
["style"]=>
   array(1) {
    ["spacing"]=>
      array(1) {
        ["blockGap"]=>
          string(21) "var:preset|spacing|20"
        }
    }
```

Some additional discussion about default style attributes in `block.json`: https://github.com/WordPress/gutenberg/discussions/53603

**Comment by onetrev** [comment]
To reduce the number of open issues, just wondering if this should actually be marked as a duplicate of this [very old issue](https://github.com/WordPress/gutenberg/issues/7342)?

Also, if this is not going to be fixed soon can we please document this bug, maybe [here](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-attributes/)? This would help prevent hours lost desperately trying to figure out why you can't get something seemingly so simple to work. Not that I did that myself. 🤪

**Comment by gziolo** [comment]
> To reduce the number of open issues, just wondering if this should actually be marked as a duplicate of this https://github.com/WordPress/gutenberg/issues/7342?

I don't think it's the same issue. In fact, I must say that #7342 is an expected behavior because the default value by design is never serialized in the block metadata.

> Make sure to include the default attributes to the "attributes" that are passed to apply_block_supports calls. I don't like this fix personally but also don't like the same fix that is present in WP_Block instance either (not sure when this was introduced) because it feels like we're hacking the code because we don't have a full "parser" in the server.

That's the fix that I'm working on right now, and it resolves the issue:

```diff
diff --git a/src/wp-includes/class-wp-block-supports.php b/src/wp-includes/class-wp-block-supports.php
index c90b5e0c54..71d6b49691 100644
--- a/src/wp-includes/class-wp-block-supports.php
+++ b/src/wp-includes/class-wp-block-supports.php
@@ -104,7 +104,7 @@ class WP_Block_Supports {
 		}
 
 		$block_attributes = array_key_exists( 'attrs', self::$block_to_render ) && is_array( self::$block_to_render['attrs'] )
-			? self::$block_to_render['attrs']
+			? $block_type->prepare_attributes_for_render( self::$block_to_render['attrs'] )
 			: array();
 
 		$output = array();
```

I agree with the sentiments shared by @youknowriad that it is suboptimal, but at the same time simple and effective. In the long run, we should rewrite the block parser. @dmsnell did some initial exploration for the lazy parser, that could absorb such operations like adding the default values automatically whenever the attribute is requested for the first time:
- https://github.com/WordPress/wordpress-develop/pull/5705

**Comment by gziolo** [comment]
Fix: https://github.com/WordPress/wordpress-develop/pull/7438
Trac Ticket: https://core.trac.wordpress.org/ticket/62114

**Comment by gziolo** [comment]
It should be resolved after WordPress 6.7 gets released in November. Related commit: https://github.com/WordPress/wordpress-develop/commit/7d0e751ffbea99583e2dda8518ff5cc915bf0bd9.

**Comment by youknowriad** [comment]
Thanks @gziolo I appreciate the fix.

**Comment by onetrev** [comment]
Amazing, thank you @gziolo. This will be very helpful (once WP 6.7 lands). 🎉
